File: agent/workflows/embabel-agent/embabel-agent-api/src/main/kotlin/com/embabel/agent/core/Retry.kt
/*
 * Copyright 2024-2026 Embabel Pty Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.embabel.agent.core

/**
 * Marker interface indicating that an operation should be retried on failure.
 *
 * Can be applied to exceptions, commands, actions, or any operation that may
 * fail transiently and should be retried according to the configured retry policy.
 *
 * Common use cases:
 * - Network timeouts
 * - Temporary service unavailability
 * - Rate limiting (429 errors)
 * - Transient database connection errors
 *
 * Example usage:
 * ```kotlin
 * import com.embabel.agent.core.Retryable
 *
 * class ApiTimeoutException(message: String)
 *     : RuntimeException(message), Retryable
 * ```
 *
 * @see NonRetryable
 * @see ActionException
 */
interface Retryable

/**
 * Marker interface indicating that an operation should NOT be retried on failure.
 *
 * Can be applied to exceptions, commands, actions, or any operation that fails
 * deterministically and will not succeed on retry.
 *
 * Common use cases:
 * - Validation errors
 * - Business rule violations
 * - Invalid input parameters
 * - Resource not found (404 errors)
 * - Authentication failures (401 errors)
 * - Authorization failures (403 errors)
 *
 * Example usage:
 * ```kotlin
 * import com.embabel.agent.core.NonRetryable
 *
 * class ValidationException(message: String)
 *     : RuntimeException(message), NonRetryable
 * ```
 *
 * @see Retryable
 * @see ActionException
 */
interface NonRetryable

File: agent/runtimes/agent-sandbox/sandbox-router/proxy/retry.go
// Copyright 2026 The Kubernetes Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package proxy

import (
	"errors"
	"net"
	"net/http"
	"syscall"
	"time"
)

// retryTransport wraps an http.RoundTripper and retries on dial-class errors.
// Retrying on dial failures is safe for any request — including POST/PUT with
// non-replayable bodies — because Go's http.Transport does not touch the
// request body until after a successful Dial. After-dial failures
// (response timeouts, mid-stream EOF, etc.) bypass the retry path so we never
// risk re-sending a partially-consumed body.
//
// Retry budget is bounded both by maxAttempts and by the per-request context
// deadline (set by the caller via context.WithTimeout). Sleeps are interrupted
// when the context fires.
type retryTransport struct {
	base         http.RoundTripper
	maxAttempts  int           // total attempts including the first; 1 = no retries
	initialDelay time.Duration // delay before the first retry
	maxDelay     time.Duration // upper bound on the per-iteration delay
	onRetry      func(req *http.Request, err error, attempt int)
}

// newRetryTransport wraps base. attempts <= 1 disables retries entirely.
func newRetryTransport(base http.RoundTripper, attempts int, initialDelay, maxDelay time.Duration, onRetry func(*http.Request, error, int)) *retryTransport {
	if attempts < 1 {
		attempts = 1
	}
	if initialDelay <= 0 {
		initialDelay = 200 * time.Millisecond
	}
	if maxDelay <= 0 {
		maxDelay = 800 * time.Millisecond
	}
	return &retryTransport{
		base:         base,
		maxAttempts:  attempts,
		initialDelay: initialDelay,
		maxDelay:     maxDelay,
		onRetry:      onRetry,
	}
}

// RoundTrip implements http.RoundTripper.
func (t *retryTransport) RoundTrip(req *http.Request) (*http.Response, error) {
	delay := t.initialDelay
	var lastErr error

	timer := time.NewTimer(delay)
	defer timer.Stop()

	for attempt := 1; attempt <= t.maxAttempts; attempt++ {
		resp, err := t.base.RoundTrip(req)
		if err == nil {
			return resp, nil
		}
		lastErr = err
		if !isRetriableDialError(err) || attempt == t.maxAttempts {
			return nil, err
		}
		if t.onRetry != nil {
			t.onRetry(req, err, attempt)
		}

		timer.Stop()
		select {
		case <-timer.C:
		default:
		}
		timer.Reset(delay)

		select {
		case <-timer.C:
		case <-req.Context().Done():
			// Surface the original dial error rather than the context error so
			// the proxy ErrorHandler reports the actual upstream failure.
			return nil, lastErr
		}

		delay *= 2
		if delay > t.maxDelay {
			delay = t.maxDelay
		}
	}
	return nil, lastErr
}

// isRetriableDialError reports whether err indicates a dial-time failure that
// is safe to retry: either DNS resolution failed (the sandbox Service may not
// be registered yet) or the TCP connect was refused / timed out (the sandbox
// Pod may not be listening yet).
//
// Errors from later in the request lifecycle (e.g. response read timeouts)
// are intentionally NOT retried because the request body may already have
// been streamed to the upstream and replaying it would duplicate side effects.
func isRetriableDialError(err error) bool {
	if err == nil {
		return false
	}
	var dnsErr *net.DNSError
	if errors.As(err, &dnsErr) {
		return true
	}
	var opErr *net.OpError
	if errors.As(err, &opErr) {
		// "dial" covers connection refused, no route, ECONNRESET on connect,
		// and dial deadline. Any other Op (e.g. "read", "write") happens
		// after the body has been touched and is intentionally not retried.
		if opErr.Op == "dial" {
			return true
		}
	}
	return false
}

// isDeadHostDialError reports whether a dial failure indicates the target
// IP itself is unreachable (timeout, no route), as opposed to a live host
// refusing the connection. A refusal (RST) proves something answers at
// that IP — e.g. the caller picked a port the Pod isn't listening on — so
// the cache entry it came from is not stale and must not be evicted.
func isDeadHostDialError(err error) bool {
	if !isRetriableDialError(err) {
		return false
	}
	return !errors.Is(err, syscall.ECONNREFUSED)
}
